### ISSUE TITLE
Clients over time popover gets cut off

### ISSUE THREAD

**Issue by WaLLy3K** [pi-hole/AdminLTE#603]
**In raising this issue, I confirm the following:**

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*
- [x] The issue I'm reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the codebase?:**

3

---

![untitled-2](https://user-images.githubusercontent.com/3049142/31572229-adb3da42-b0ec-11e7-93ed-d575a5298e18.png)

![untitled-3](https://user-images.githubusercontent.com/3049142/31572247-115842cc-b0ed-11e7-8888-918485bc31b5.png)

The new Clients over Time popover gets cut off, rendering it basically unusable for determining which client made a huge spike in queries.

**Comment by andreskrey** [comment]
Same thing happens with the "Forward Destinations over Time" when you have multiple DNS servers.

**Comment by PromoFaux** [comment]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/AdminLTE/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/AdminLTE/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`10`

---
**Expected behaviour:**

Tooltips should render outside of the canvas

**Actual behaviour:**

Tooltips only render within the canvas.

**Steps to reproduce:**

Mouseover a chart, and look at it. 

**Troubleshooting undertaken, and/or other relevant information:**

A possible fix is to use [External (Custom) Tooltip](http://www.chartjs.org/docs/latest/configuration/tooltip.html?h=external%20)

Some discussion on it here : https://github.com/chartjs/Chart.js/issues/622

If I can get some time, I'll take a look into it.

**Comment by DL6ER** [comment]
[This article on Discourse](https://discourse.pi-hole.net/t/web-ui-clients-over-time-overlay-tooltip-cut-off-on-chart-workaround/6361) is suggesting to increase the height of the graphs notable to mitigate this issue, what do you think @pi-hole/admin ?

**Comment by PromoFaux** [comment]
I mean, yeah, that's one way of doing it, but I think it would be nicer to go down the custom tooltip path. That said, I've had no time to work on that myself lately!

**Comment by Mcat12** [comment]
With enough clients, it will still be an issue with that workaround.

**Comment by excyruss** [comment]
It helps, but I you get repeating similar colours too (which is probably another issue, sorry).

![pihole](https://user-images.githubusercontent.com/28961015/34733843-b01039f2-f561-11e7-8bf1-7b53702e4857.jpg)

We need fewer devices in the home but I doubt that's the trend :)

**Comment by borpin** [comment]
Could the position of the legend be made static on the RHS of the chart, but the data fed from where the mouse is in terms of updating the legend?

**Comment by Mcat12** [comment]
I haven't check that case, but it would depend on if the library supports it. I would guess not because it renders everything in a canvas

**Comment by borpin** [comment]
Couple of thoughts:

- The 'Queries over last 24 hours' chart tooltip appears on top or to left of mouse location which is better.  
- The 'Clients (over time)' chart it always is displayed below the mouse location. If it too was to the left or above, that would help.

On the 'Clients (over time)' chart, removing those clients with no data would help as well.

An alternative approach could be to make the title of the chart clickable, taking the user to the chart on a single page, the chart could be much bigger (with more canvas space below the axis which might help.

Could something like this be done? http://hoboman313.github.io/chartsjs-demo/

This http://www.flotcharts.org/flot/examples/tracking/index.html is quite a neat solution with a fixed legend though it uses flot and I have not been able to find a chartjs solution to match it.

**Comment by maciboy** [comment]
@excyruss 
The issue with too many clients can be diminished by filtering all clients with currently 0 requests.
In your case only 6 clients would be shown then…
Of course this is not a full solution, but with the assumption that not all clients are active in every timespan, it would at least improve the situation 😄

**Comment by kh0505** [comment]
@maciboy 
Thanks. Updating that section of code strips out the 0 count clients which I don't care about anyway. It's not a perfect solution but makes most areas on my graph not cut off the tooltip as I have around 18 clients. This has been bugging me for quite a while. I have researched this and other applications using Chart.js have the same issue. As PromoFaux suggested above it appears that a Custom Tooltip is the only real solution. My coding skills are a little to rusty for that these days but for now this solution will suffice. 

Easy enough for anyone to update on their own. 

                        label: function(tooltipItems, data) {
                                if(tooltipItems.yLabel != 0) {
                                    return data.datasets[tooltipItems.datasetIndex].label + ": " + tooltipItems.yLabel;
                                }
                        }

**Comment by kh0505** [comment]
This is how to get a customtooltip that renders outside of the clients over time graph. I didn't apply it to the queries over time graph. Also, if you implemented the stripping out of 0 count clients above I would suggest undoing that as you will end up with blank lines with this new code. I am going to enjoy the rest of my Saturday rather than fiddling with that right now...  The below is pretty much a hack of me googling and researching but it works. Not as pretty as it doesn't have the triangle fly away on the tooltip but it works. I am sure there is some process to officially submit it on github but I don't know if I am going to work on this anymore since I got it to do what I wanted. So do what you will with it. 

[piholetooltipfix.txt](https://github.com/pi-hole/AdminLTE/files/1821833/piholetooltipfix.txt)

**Comment by maciboy** [comment]
Would love to see some kind of adoption of @kh0505 modification!

Maybe you can share a screenshot, of how your custom tooltip looks in usage?

**Comment by kh0505** [comment]
![image](https://user-images.githubusercontent.com/37278546/37558200-c14bcaa6-29e6-11e8-8477-73cec394ded4.png)

**Comment by PromoFaux** [comment]
Good stuff @kh0505! I think this is one of those things we've just been putting off and off! You're free to submit a pull request against the repository (`devel` branch, please) if you'd like to.

**Comment by maciboy** [comment]
I managed to edit both files the way @kh0505 explained, for those of you wanting to test it:
[Tooltip.zip](https://github.com/pi-hole/AdminLTE/files/1822881/Tooltip.zip)

**Comment by 91ajames** [comment]
Thank you so much, hope this gets added in the next update. honestly, they should put this in now, right away. but of course some possible breaks may endure and if any security risk but I doubt it.

**Comment by PromoFaux** [comment]
We'll add it in, just giving @kh0505 a grace period should he wish to submit it himself. However, we will now add it in and credit him in the commit message

**Comment by 91ajames** [comment]
I tried getting the manual fixes and so i thought i had screwed up, or perhaps i didn't reboot my device when making the manual changes. but when u get to the Now instead of the pass few hrs, there's this bug.
https://imgur.com/a/lbgjL 2 pics

**Comment by Mcat12** [comment]
@91ajames that issue is already noted in the PR.

**Comment by 91ajames** [comment]
was just to come back and edit that its fix.

**Comment by thewraith2008** [comment]
The issues I see when using fix by kh0505. Sorry for image overload.

Tooltip window varys position depending on "Client Over Time" position in browser window.
3 images below show same position in graph but different positions in browser.
About middle of browser:
![pihole_clientsovertime_postion_top](https://user-images.githubusercontent.com/25628886/38159190-cd71d720-34e6-11e8-8c58-a12186fbcd0a.jpg)

Closer to top of browser:
![pihole_clientsovertime_postion_mid](https://user-images.githubusercontent.com/25628886/38159197-e1639cc8-34e6-11e8-9d46-b6288ae92cb5.jpg)

Top of browser (tooltip not seen):
![pihole_clientsovertime_postion_mid2](https://user-images.githubusercontent.com/25628886/38159201-e8d61cd8-34e6-11e8-99f5-6bd1d7c995fd.jpg)

Tooltip window is cut off when pointing to data on most left side of graph.

![pihole_clientsovertime_cutoff](https://user-images.githubusercontent.com/25628886/38159228-8a4df16c-34e7-11e8-9f41-04b61fb7f278.jpg)


When browser is windowed, tooltip window is cut off on left side (as above) of graph and becomes squashed as right side of graph is reached.

Notice tooltip becoming squashed:
![pihole_clientsovertime_windowed_position_bottom_brow_squashed1](https://user-images.githubusercontent.com/25628886/38159252-f60ebf08-34e7-11e8-9df4-310dce094080.jpg)

Squashed:
![pihole_clientsovertime_windowed_position_bottom_brow_squashed2](https://user-images.githubusercontent.com/25628886/38159254-fe15181e-34e7-11e8-81fc-aa155b994f61.jpg)

When browser is windowed and "Client Over Time" is positioned from top to about centre in browser, tooltip window is not visible.

When browser is windowed, tooltip is way above "Client Over Time"  and appears over "Queries over last 24 hours"

![pihole_clientsovertime_windowed_position_bottom_brow](https://user-images.githubusercontent.com/25628886/38159280-7b8595da-34e8-11e8-97aa-81be87500cdb.jpg)

The real advantage is the improvement by showing clients in order of queries.
Not sure if color duplicates are still a issue.

Thank for listening.

**Comment by DL6ER** [comment]
> and becomes squashed as right side of graph is reached.

This does not exist in the version I coded.

> Not sure if color duplicates are still a issue.

Yes, the used color scheme does only have a finite number of colors.

**Comment by thewraith2008** [comment]
> This does not exist in the version I coded.

I only see this happening when browser is windowed. That is not maximized/fullscreen.
I see same on Firefox 59.0.2 and Internet Explorer 11.0.9600.18860 (Win7 32bit)

I'm running v3.2.1 AdminLTE. I have been reluctant to update to v3.3 after problems have been reported with it.

**Comment by ScottHelme** [comment]
I came to raise the same issue and found this one. I wrote a [blog](https://scotthelme.co.uk/securing-dns-across-all-of-my-devices-with-pihole-dns-over-https-1-1-1-1/) on how to setup Pi-Hole on a Raspberry Pi and I've had a few people point this issue out to me. Would be awesome to see a fix as this is one of the main graphs on the dashboard and the first thing you see.

![image](https://user-images.githubusercontent.com/5352840/38729806-0ddd2cd6-3f0c-11e8-86d0-c8378930e2c9.png)

**Comment by thewraith2008** [comment]
The issue can be mitigated by applying by hand the changes to the 2 files (plain text) indicated in the above posts. This is a low priority item that the DEV team will get to one day, for now doing it for yourself will be the only option if you really need something. Also if you make the changes, you can give feedback to assist with its improvement.

**Comment by PromoFaux** [comment]
It's in progress:
https://github.com/pi-hole/AdminLTE/pull/700

**Comment by 91ajames** [comment]
was working for a while until last week or so, it reverted back to half screen unseeable adress's.

**Comment by thewraith2008** [comment]
@91ajames if you have done an Update (pihole -up) or Reconfigure/Repair (pihole -r) of Pihole in the last week or so, this will overwrite (restore to release default) any manual changes that where made by hand.

**Comment by Mcat12** [comment]
The PR has been merged into development

**Comment by bryceacc** [comment]
Sorry for the stupid question but how do I modify these files to implement this fix myself?  I stopped lighttpd but the files are still not writable and even if I force a write and then restart lighttpd or the entire pi, the changes don't actually get reflected in the admin page once it starts up again.  How do I modify the files in /var/www/html/admin/ ?

**Comment by Mcat12** [comment]
I would not suggest modifying the files yourself. They are writable if you use `sudo`.

**Comment by bryceacc** [comment]
Well that's the thing, I transferred the modified files and `sudo cp`'ed them into the right places. `cat `on the index.js and .php file both show the correct modifications but after a restart, the admin console is not fixed.  Chrome tools show the source code and they are unmodified.  Is there a different directory that the web server is running from?  

I modified these two files to kh0505's instruction:

/var/www/html/admin/scripts/pi-hole/js/index.js
/var/www/html/admin/index.php

**Comment by dschaper** [comment]
Sounds like a bad SDCard. Try touching a file in the home directory, rebooting and see if that file still exists after the reboot.

**Comment by bryceacc** [comment]
touch test is fine, I also `cat `checked the modifications before and after reboot so I know they are written.  Unless `sudo reboot` isn't performing a complete restart?  

Everything else about the pi and pihole has been amazing, I just turned it into the DHCP server though and that grew my client list to the correct amount of clients I have around which is way too many for the small graph.  Even just putting in the 0 requests check helps a lot, and I can modify that from chrome tools but it would be nice to be a persistent change

**Comment by bryceacc** [comment]
Correct, cat on the two index files shows the changes at all times after multiple reboots, I look at the source displayed in my browser and it's the original unmodified code that I wanted changed.

> pi@raspberrypi:~ $ sudo systemctl --no-pager --full lighttpd.service
> Unknown operation lighttpd.service.

The other command looks like a good ol config file but messes up markdown so I won't insert that.  The parts that specify directories and files are:

server.document-root        = "/var/www/html"
server.error-handler-404    = "pihole/index.php"
server.upload-dirs          = ( "/var/cache/lighttpd/uploads" )
server.errorlog             = "/var/log/lighttpd/error.log"
server.pid-file             = "/var/run/lighttpd.pid"
server.username             = "www-data"
server.groupname            = "www-data"
server.port                 = 80
accesslog.filename          = "/var/log/lighttpd/access.log"
accesslog.format            = "%{%s}t|%V|%r|%s|%b"
index-file.names            = ( "index.php", "index.html", "index.lighttpd.html" )

**Comment by dschaper** [comment]
It's `sudo systemctl --no-pager --full status lighttpd.service` I dropped a word. And you can paste in the full config and wrap the text with ` ``` ` backticks to preserve the original formatting.

**Comment by dschaper** [comment]
Also, may not make a difference but have you `CTRL-F5` in the browser to hard clear any cache?

**Comment by bryceacc** [comment]
oh no... 🤦 hard refresh works... thanks for your help!

**Comment by DL6ER** [comment]
@Melantrix Please try a full refresh (`Ctrl + F5` or `Shift + Ctrl + R`). Your browser might still be caching the old version of the scripts.

**Comment by Melantrix** [comment]
I've done that but after manually deleting chrome and reinstalling it. (NB: I am on mobile; Chrome on Android)

I've just tried it on a desktop and there it now works correctly. Is there

**Comment by DL6ER** [comment]
@Melantrix Do you mean that it still doesn't work?

**Comment by Melantrix** [comment]
Oh I'm sorry, that message was absolutely unclear xD it works now, but only after reinstalling the app..
